File: sdks/python/salanor_aegis/events.py
"""Event builders for policy and HTTP result records."""

from __future__ import annotations

from datetime import datetime, timezone
from typing import Any, Literal

from salanor_aegis.ids import new_event_id

PolicyDecision = Literal["allow", "deny", "allow_with_obligation", "allow_retro_audit"]


def _utc_now_iso() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z"


def build_policy_decision_event(
    *,
    organization_id: str,
    trace_id: str,
    agent_id: str,
    key_id: str,
    tool_name: str,
    decision: PolicyDecision,
    actor_principal: str,
    policy_id: str | None = None,
    rule_id: str | None = None,
    reason: str | None = None,
    span_id: str | None = None,
    parent_span_id: str | None = None,
    payload_extras: dict[str, Any] | None = None,
) -> dict[str, Any]:
    denied = decision == "deny"
    payload: dict[str, Any] = {
        "policy_id": policy_id,
        "rule_id": rule_id or ("deny:unattributed" if denied else "allow:default"),
        "reason": reason or ("policy denied" if denied else "policy allowed"),
        **(payload_extras or {}),
    }
    event: dict[str, Any] = {
        "schema_version": 1,
        "event_id": new_event_id(),
        "organization_id": organization_id,
        "trace_id": trace_id,
        "agent_id": agent_id,
        "key_id": key_id,
        "emitted_at": _utc_now_iso(),
        "actor_type": "agent",
        "actor_principal": actor_principal,
        "action_kind": "policy_decision",
        "policy_decision": decision,
        "tool_name": tool_name,
        "payload": payload,
    }
    if policy_id and policy_id != "none":
        event["policy_id"] = policy_id
    if span_id:
        event["span_id"] = span_id
    if parent_span_id:
        event["parent_span_id"] = parent_span_id
    return event

File: sdks/typescript/src/proxy/events.ts
import { randomUUID } from "node:crypto";
import type { ApsEvent, PolicyDecision } from "../types.js";

export function newEventId(): string {
  return `evt_${randomUUID().replace(/-/g, "").slice(0, 24)}`;
}

export function buildPolicyDecisionEvent(params: {
  organizationId: string;
  traceId: string;
  agentId: string;
  keyId: string;
  toolName: string;
  decision: PolicyDecision;
  actorPrincipal: string;
  policyId?: string | null;
  ruleId?: string | null;
  reason?: string;
  payloadExtras?: Record<string, unknown>;
}): ApsEvent {
  const denied = params.decision === "deny";
  return {
    schema_version: 1,
    event_id: newEventId(),
    organization_id: params.organizationId,
    trace_id: params.traceId,
    agent_id: params.agentId,
    key_id: params.keyId,
    emitted_at: new Date().toISOString(),
    actor_type: "agent",
    actor_principal: params.actorPrincipal,
    action_kind: "policy_decision",
    policy_decision: params.decision,
    tool_name: params.toolName,
    policy_id:
      params.policyId && params.policyId !== "none" ? params.policyId : undefined,
    payload: {
      policy_id: params.policyId ?? null,
      rule_id:
        params.ruleId ??
        (denied ? "deny:unattributed" : "allow:default"),
      reason:
        params.reason ??
        (denied ? "policy denied" : "policy allowed"),
      ...params.payloadExtras,
    },
  };
}

export function httpResultStatus(httpStatus: number): "ok" | "error" | "timeout" | "blocked" {
  if (httpStatus >= 200 && httpStatus < 400) {
    return "ok";
  }
  if (httpStatus === 408 || httpStatus === 504) {
    return "timeout";
  }
  return "error";
}

export function buildResultEvent(params: {
  organizationId: string;
  traceId: string;
  agentId: string;
  keyId: string;
  toolName: string;
  actorPrincipal: string;
  parentEventId: string;
  httpStatus: number;
}): ApsEvent {
  const resultStatus = httpResultStatus(params.httpStatus);
  return {
    schema_version: 1,
    event_id: newEventId(),
    organization_id: params.organizationId,
    trace_id: params.traceId,
    agent_id: params.agentId,
    key_id: params.keyId,
    emitted_at: new Date().toISOString(),
    actor_type: "agent",
    actor_principal: params.actorPrincipal,
    action_kind: "result",
    policy_decision: "allow",
    tool_name: params.toolName,
    parent_event_id: params.parentEventId,
    result_status: resultStatus,
    payload: { http_status: params.httpStatus, result_status: resultStatus },
  };
}
